diffrax/integrate.py

```diff
@@ -178,23 +236,76 @@ def _apply_history(x):
 
         # TODO: handle discontinuity detection for delays
         error_order = solver.error_order(terms)
+
         (
             keep_step,
-            tprev,
-            tnext,
+            tprev_candidate,
+            tnext_candidate,
             made_jump,
             controller_state,
             stepsize_controller_result,
         ) = stepsize_controller.adapt_step_size(
-            state.tprev,
-            state.tnext,
+            tprev,
+            tnext,
             state.y,
             y,
             args,
             y_error,
             error_order,
             state.controller_state,
         )
+
+        # Finding all of the potential discontinuity roots
+        if delays is not None:
+            roots = event_function_root_finding(
+                delays,
+                state.discontinuity,
+                tprev,
+                tnext,
+                state.y,
+                y,
+            )
+            # Get bool mask array of the only valid ones
+            new_discont_bool = (
+                (roots > state.discontinuity[state.discontinuity_save_index - 1])
+                & (roots < tnext)
+                & (roots > tprev)
+            )
+
+            # Wacky way of finding new potential roots can definitely be a smarter way
+            proxy_best_candidate_finder = (
+                jnp.invert(new_discont_bool) * roots + new_discont_bool * roots
+            )
+
+            best_candidate = jax.lax.cond(
+                jnp.invert(new_discont_bool).all(),
+                lambda _: jnp.nan,
+                lambda _: jnp.nanmin(proxy_best_candidate_finder),
+                None,
+            )
+
+            if isinstance(stepsize_controller, AbstractAdaptiveStepSizeController):
+                (discont_keep_step, tprev, tnext, discont_made_jump,) = jax.lax.cond(
+                    best_candidate > 0,
+                    lambda _: (False, state.tprev, best_candidate, True),
+                    lambda _: (True, tprev_candidate, tnext_candidate, False),
+                    None,
+                )
+            else:
+                (discont_keep_step, tprev, tnext, discont_made_jump,) = jax.lax.cond(
+                    best_candidate > 0,
+                    lambda _: (False, state.tprev, best_candidate, True),
+                    lambda _: (True, tprev_candidate, tnext_candidate, False),
+                    None,
+                )
+
+            # this conditional enables to handle edge cases where t1 is
+            # simultaneously a discontinuity and the end of integration
+
+            pred = jnp.isclose(t1, tnext)
+            discont_keep_step = jnp.where(pred, True, discont_keep_step)
+            keep_step = keep_step & discont_keep_step
+
```

Comment:
I think all of the above logic can be made a lot faster. (Which is likely the source of much of the slowdowns you're seeing.)

The logic is a bit complicated to write out efficiently. Something like the following, I think. (Untested.)

```python
def _maybe_find_discontinuity():
    dense_ts = jnp.stack([tprev, tnext])
    dense_infos = jtu.tree_map(lambda x: x[None], dense_info)
    dense_discont = DenseInterpolation(
        ts=dense_ts,
        ts_size=2,
        interpolation_cls=solver.interpolation_cls,
        infos=dense_infos,
        direction=1,
    )
    flat_delays = jtu.tree_leaves(delays)
    gs = []
    for delay in flat_delays:
        g = lambda t, delay=delay: t - delay(t, dense_discont.evaluate(t), args) - state.discontinuities
        gs.append(g)
        
    def _find_discontinuity():
        # Start by doing a cheap bisection search to reduce over the stored-discontinuity dimension.
        
        def _cond_fun(_val):
            _, _, _pred = val
            return _pred

        def _body_fun(_val):
            _ta, _tb = val
            _tmid = _ta + 0.5 * (_tb - _ta)
            _gas = jnp.stack([jnp.sign(g(_ta)) for g in _gs])
            _gmids = jnp.stack([jnp.sign(g(_tmid)) for g in _gs])
            _gbs = jnp.stack([jnp.sign(g(_tb)) for g in _gs])
            _any_left = jnp.any(_gas != _gmid)
            _next_ta = jnp.where(_any_left, _ta, _tmid)
            _next_tb = jnp.where(_any_left, _tmid, _tb)
            _pred = jnp.any(jnp.sum(_gas != _gbs, axis=1) > 1)
            return _next_ta, _next_tb, _pred
            
        _init_val = (tprev, tnext, True)
        _final_val = lax.while_loop(_cond_fun, _body_fun, _init_val)
        _, _tb, _ = final_val
        
        # Then do a more expensive Newton search to find the first discontinuity.
        _discont_solver = NewtonNonlinearSolver(rtol=..., atol=...)
        _disconts = []
        for g, delay in zip(gs, flat_delays):
            _i = jnp.argmax(jnp.sign(g(_ta)) != jnp.sign(g(_tb)))
            _d = state.discontinuities[_i]
            _h = lambda t, delay=delay, _d=_d: t - delay(t, dense_discont.evaluate(t), args) - _d 
            _discont = _discont_solver(_h, _tb, args).root
            _disconts.append(_discont)
        return jnp.min(jnp.stack(_disconts))
            
    def _find_discontinuity_wrapper():
        return lax.cond(jnp.any(init_discont), _find_discontinuity, lambda: tnext)
        
    init_discont = jnp.stack([jnp.sign(g(tprev)) != jnp.sign(g(tnext)) for g in gs])
    # We might have rejected the step for normal reasons; skip looking for a discontinuity if so.
    return lax.cond(unvmap_any(init_discont), _find_discontinuity_wrapper, lambda: tnext)

def _maybe_find_discontinuity_wrapper():
    # Only look for a discontinuity on rejected steps.
    return lax.cond(keep_step, lambda: tnext, _maybe_find_discontinuity)

tnext = lax.cond(unvmap_all(keep_step), lambda: tnext, _maybe_find_discontinuity_wrapper)
```
At the moment that's all written out as if it was inline code; it may be a bit more readable to split this out into some helper functions. I invite you to rewrite/rename something like this however you feel is neatest; I wrote it rather hastily.

This should be much more efficient. We:
1. Only search for a discontinuity if the step is rejected.
2. Only need to update `tnext`, and not anything else.
3. Use a cheap bisection search to identify the single discontinuity we need to do a Newton search for. (For each delay function.)
4. `vmap` efficiently: skip searching for a discontinuity if none of our batch elements hit a discontinuity.
    - This is what's going on with the `unvmap` wrappers. As applied to a single batch element, then `unvmap_{any,all}` act exactly the same as `jnp.{any,all}`, so the wrappers in the above code look a bit like they do nothing. The magic is that `unvmap_{any,all}` will ignore any `jax.vmap`s around the entire code, and will do an `any/all` operation over the whole thing. This means we can skip these computations if *none* of our batch elements hit them. In contrast if we were to leave these off, then by default `jax.vmap` of `lax.cond` will instead take the rather naive approach of always evaluating both branches, and then merging them together.

(Point 4 is pretty advanced JAX.)

Incidentally I definitely don't claim that the above implementation is optimally efficient. Like I said; hastily thrown together and I may have overlooked something. Mostly I'm hoping that doing a bisection search over the previous discontinuities will make the Newton step that much more efficient, since we will have many fewer Newton solvers to do.

---

Various other comments on the above (that might not matter once it's rewritten):
1. Both branches of the `isinstance` check look to be the same.
2. We shouldn't ever need to modify `keep_step`. We only ever trigger discontinuity handling on rejected steps.
    - After all, some discontinuities may be minor -- e.g. it may be the case that `y0_history` extends from our solution smoothly, and we have no discontinuities at all.
3. Likewise, we shouldn't ever need to update `tprev`. (This will have been done for us in the step size controller on a rejected step.)
4. Prefer `lax.cond` over `jax.lax.cond`.
5. We don't need to do `lax.cond(pred, lambda _: foo lambda _: bar, None)` any more; just `lax.cond(pred, lambda: foo, lambda: bar)` will suffice. (Some of the existing Diffrax code was written before this was possible.)
6. I don't think your inputs into `event_function_root_finding` are consistent. `state.y` is the value at `state.tprev`

Reply:
Thanks, i'll definitely take a look and see how we can incorporate that !  One thing that stands out is the fact that we can't use functions as operands for `lax.cond` and passing `delays` here would be a bottleneck. I'll look a bit more in the `unvmap` looks fancy.

Reply:
Regarding 2/, I have thought about it and im not sure that we cant leave `keep_step` be. For example, for a `ConstantController`, `keep_step` is always `True` ? (and we wouldn't be able to handle discontinuities if we do it with `keep_step`).In that regards, I don't know how you would want me to handle the discontinuity jumps. Moreover, i have yet to find a justification that proves that if the step is rejected then we have a discontinuity. I would rather thing the other way around ie that regarless of if the step is accepted or not we might find a discontinuity.

Reply:
https://github.com/patrick-kidger/diffrax/pull/169#discussion_r992431764. If I look at jax documentation for `lax.cond` the operand type must be PyTree (list, array or tuple) (https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html?highlight=cond%20#jax.lax.cond). And having a `delays` arguments of a `PyTree[Callable]` will throw an error that ressembles : 
`TypeError: Value <function <lambda> at 0x7f86dc175430> with type <class 'function'> is not a valid JAX type`.

Reply:
Regarding `keep_step`. I feel confident that we should only search for a discontinuity if we reject a step. (Indeed this is what DelayDiffEq.jl does too.) It's a very classic "problem" that discontinuities lead to step rejections; this comes up frequently when solving diffeqs with a piecewise control. Discontinuities usually produce quite dramatic changes in the size of the vector field, and any step made over this will incur a large error, leading to the step rejection.

And of course computationally, it's great if we can skip doing the expensive discontinuity search whenever possible.

Regarding `ConstantStepSize`: this is a good point, that this fixes `keep_step=True` and that this would mean we never search for a discontinuity. On balance, I think I'm happy with that. It means we really will always be taking constant step sizes.

---

Regarding `lax.cond`: where in my code snippet do I pass `delays` as an argument to `lax.cond`? I agree that wouldn't work, but I don't think I do this.

Reply:
This makes sense to check for jumps if the step was rejected as you said for computational reasons. I suppose that in the case the discontinuities propagate smoothly in the DDE this approach might fail. We can go with this approach then if you are convinced! 

For the 'ConstantController' im not sure to understand you want to leave it as is ? 

## 

To make the code more modular by putting it in seperate files we would have to call each 'lax.cond' with their arguments (including 'delays,) no ? Im probably not familar enough with just the 'lax.cond, passing no arguments

Reply:
Yup, I think we should just leave `ConstantStepSize` alone; don't do any discontinuity handling with it.

Regarding `lax.cond`, notice how the approach I'm taking above is just to capture everything by closure. Imagine that we just hardcoded the body of each delay function in; as far as JAX is concerned this is the same as writing that function elsewhere and then calling it. (The latter being what we're doing.) Since JAX operates by tracing it can't tell the difference.

Reply:
Sounds good ill try to put/modify the discontinuity handling in 'integrate.py' then. 

By choosing to not touch the constant controller  DDEs will only be compatible with PIDController.